trin-core/src/portalnet/types/metric.rs

```diff
@@ -45,6 +96,41 @@ mod test {
         }
     }
 
+    #[test]
+    fn distance_log2() {
+        fn prop(x: DhtPoint) -> TestResult {
+            let x = U256::from_big_endian(&x.0);
+            let distance = Distance(x);
+            let log2_distance = distance.log2();
+
+            match log2_distance {
+                Some(log2) => {
+                    let x_floor = U256::from(1u8) << log2 - 1;
+                    let x_ceil = if log2 == 256 {
+                        U256::MAX
+                    } else {
+                        U256::from(1u8) << log2
+                    };
+
+                    TestResult::from_bool(x >= x_floor && x <= x_ceil)
```

Comment:
I think it would be a correct tighter bound to say that x needs to be strictly beneath the ceiling. In order to handle the 256 case, it could follow this approach:
```suggestion
                    if log2 == 256 {
                        TestResult::from_bool(x >= x_floor && x <= U256::MAX)
                    } else {
                        let x_ceil = U256::from(1u8) << log2;
                        TestResult::from_bool(x >= x_floor && x < x_ceil)
                    }
```

The MAX check is actually a little redundant, since U256 cannot be higher, so it would also be fine to just use this in the 256 case IMO: `TestResult::from_bool(x >= x_floor)`